inc/azure_macro_utils/macro_utils.h

```diff
@@ -202,6 +202,61 @@ const char* MU_C3(MU_, enumIdentifier,_ToString)(enumIdentifier value)
 
 #define MU_ENUM_VALUE_2(enumIdentifier, value) "", MU_ENUM_TO_STRING_2(enumIdentifier, (value)), (int)(value)
 
+#define CONVERT_ENUM_VALUE(from, to) \
+    case (from): \
+        *enum_value_to = (to); \
+        return 0; \
+
+// This macro declares an enum conversion function that translates a value from one enum to another
+#define MU_DECLARE_CONVERT_ENUM(ENUM_TYPE_FROM, ENUM_TYPE_TO) \
+    int MU_C4(convert_enum_, ENUM_TYPE_FROM, _, ENUM_TYPE_TO)(ENUM_TYPE_FROM enum_value_from, ENUM_TYPE_TO* enum_value_to);
+
+// This macro defines an enum conversion function that translates a value from one enum to another
+// It is implemented as a switch statement. This allows for catching multiple from values being in the list also
+#define MU_DEFINE_CONVERT_ENUM_WITHOUT_INVALID(ENUM_TYPE_FROM, ENUM_TYPE_TO, ...) \
+    int MU_C4(convert_enum_, ENUM_TYPE_FROM, _, ENUM_TYPE_TO)(ENUM_TYPE_FROM enum_value_from, ENUM_TYPE_TO* enum_value_to) \
+    { \
+        switch (enum_value_from) \
+        { \
+            default: \
+                break; \
```

Comment:
I wonder if it's worth having a version that directly returns a value and has a default defined in the macro (which could just wrap this even)? Just makes the calling code simpler and I can't imagine many places where we want to handle failures for this.


This seems like it could easily be hidden in the macro (or another version):

```c
OTHER_ENUM result;
if (MU_CONVERT_ENUM(SOME_ENUM, OTHER_ENUM)(something, &result) != 0)
{
    // maybe LogError
    result = OTHER_ENUM_INVALID;
}
````

Reply:
I think either way the caller will have to check if an error happened. So checking against a predefined enum value that is added to the enum (like _INVALID) or checking against the return of the function makes not that much of a difference as the if still has to be written, right?

---
In reply to: [394481869](https://github.com/Azure/azure-macro-utils-c/pull/36#discussion_r394481869) [](ancestors = 394481869)